sound/soc/sof/intel/telemetry.h

```diff
@@ -0,0 +1,207 @@
+ */
+
+#ifndef _SOF_INTEL_TELEMETRY_H
+#define _SOF_INTEL_TELEMETRY_H
+
+#define SOF_IPC4_FW_REPORTED_MAX_CORES_COUNT	5
+#define SOF_IPC4_FW_LP_TIMER_DELAY_CAUSE_COUNT	2
+#define SOF_IPC4_FW_MAX_CONDITION_DATA_COUNT	9
+#define SOF_IPC4_FW_MAX_NUMBER_MEM_BINS		1
+#define SOF_IPC4_FW_MAX_REPORTED_TASKS		40
+#define SOF_IPC4_FW_AREGS_COUNT			64
+
+struct system_tick_info_s {
+	u32 count;
+	u32 last_sys_tick_count;
+	u32 max_sys_tick_count;
+	u32 last_ccount;
+	u32 avg_utilization;
+	u32 peak_utilization;
+	u32 peak_utilization_4k;
+	u32 peak_utilization_8k;
+	u32 rsvd[2];
+};
+
+struct deadlock_info_s {
+	u32 register_a0;
+	u32 register_a1;
+	u32 cached_stack_ptr;
+	/* caller, params, condition and return value  for deadlock */
+	u32 run_condition_data[SOF_IPC4_FW_MAX_CONDITION_DATA_COUNT];
+};
+
+struct assert_statistics_s {
+	u32 asserts_triggered_count;
+	u32 last_assert_log_id;
+};
+
+struct assert_info_s {
+	u32 mode;
+	struct assert_statistics_s stats[SOF_IPC4_FW_REPORTED_MAX_CORES_COUNT];
+};
+
+struct xxxruns_info_s {
+	u32 gateway_underruns_count;
+	u32 last_underrun_gateway_node_id;
+	u32 gateway_overruns_count;
+	u32 last_overrun_gateway_node_id;
+	u32 mixer_underruns_count;
+	u32 last_mixer_underrun_resource_id;
+};
+
+struct performance_info_s {
+	u32 cpu_frequency;
+	u32 dropped_lp_timer_ticks_cause[SOF_IPC4_FW_LP_TIMER_DELAY_CAUSE_COUNT];
+};
+
+struct sram_mem_info_s {
+	u32 total_pages_cnt;
+	u32 pages_in_use_cnt;
+	u32 phys_pages_in_use_cnt;
+};
+
+struct sram_mem_pool_info_s {
+	u32 registered_pools_cnt;
+	u32 pools_in_use_cnt;
+	u32 registered_pages_cnt;
+	u32 pages_in_use_cnt;
+};
+
+/* Global information about SRAM opportunistic memory pools usage */
+struct sram_opportunistic_mem_pool_info_s {
+	u32 registered_opportunistic_pools_cnt;
+	u32 opportunistic_pools_in_use_cnt;
+	u32 registered_opportunistic_pages_cnt;
+	u32 opportunistic_pages_in_use_cnt;
+	u32 requested_opportunistic_pages_cnt;
+};
+
+struct mem_pools_info_s {
+	struct sram_mem_info_s hpsram_mem_info;
+	struct sram_mem_pool_info_s hpsram_mem_pool_info;
+	struct sram_opportunistic_mem_pool_info_s hpsram_opportunistic_pool_info;
+	struct sram_mem_info_s lpsram_mem_info;
+	struct sram_mem_pool_info_s lpsram_mem_pool_info;
+	struct sram_opportunistic_mem_pool_info_s lpsram_opportunistic_pool_info;
+};
+
+struct timeout_info_s {
+	u32 expected_value;
+	u32 current_value;
+};
+
+/* obstacles for ultra low power entry */
+struct ulp_prevention_info_s {
+	u8 prevented_by_high_power;
+	u8 prevented_by_idle_mode_count_low;
+	u8 idle_mode_count_down_;
+	u8 prevented_by_on_pipeline_state;
+	u8 prevented_by_fastmode_task;
+	u8 prevent_autonomous_reset_counter;
+	u8 rsvd[2];
+};
+
+struct transition_info_s {
+	u32 entry_cnt;
+	u32 exit_cnt;
+	u32 last_entry_systick;
+};
+
+/* parameters related to WHM Ultra Low Power mode */
+struct ulp_telemetry_s {
+	struct ulp_prevention_info_s	prevention_info;
+	struct transition_info_s	transition_info;
+	u32				data[4];
+};
```

Comment:
do we really case about all these ULP things?

I really question the ability even for an expert developer to use anything above this line.

I also wonder if anyone at Intel is really ready to freeze these structures as "THE" Windows/Linux ABI for telemetry. I just don't see how it might scale or be extended to new modules.

Reply:
remove these unused ULP things and use a opaque array of u32 instead